pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/BucketDelayedDeliveryTracker.java

```diff
@@ -0,0 +1,569 @@
+                bucketState.getSnapshotCreateFuture().orElseGet(() -> CompletableFuture.completedFuture(null))
+                        .thenApply(__ -> null);
+
+        return snapshotCreateFuture.thenCompose(__ -> {
+            final long bucketId = getBucketId(bucketState);
+            CompletableFuture<Integer> loadMetaDataFuture = new CompletableFuture<>();
+            if (isRecover) {
+                // TODO Recover bucket snapshot
+            } else {
+                loadMetaDataFuture.complete(bucketState.currentSegmentEntryId + 1);
+            }
+
+            return loadMetaDataFuture.thenCompose(nextSegmentEntryId -> {
+                if (nextSegmentEntryId > bucketState.lastSegmentEntryId) {
+                    // TODO Delete bucket snapshot
+                    return CompletableFuture.completedFuture(null);
+                }
+
+                return bucketSnapshotStorage.getBucketSnapshotSegment(bucketId, nextSegmentEntryId, nextSegmentEntryId)
+                        .thenAccept(bucketSnapshotSegments -> {
+                            if (CollectionUtils.isEmpty(bucketSnapshotSegments)) {
+                                return;
+                            }
+
+                            SnapshotSegment snapshotSegment = bucketSnapshotSegments.get(0);
+                            List<DelayedIndex> indexList = snapshotSegment.getIndexesList();
+                            DelayedIndex lastDelayedIndex = indexList.get(indexList.size() - 1);
+
+                            this.snapshotSegmentLastIndexTable.put(lastDelayedIndex.getLedgerId(),
+                                    lastDelayedIndex.getEntryId(), bucketState);
+
+                            for (DelayedIndex index : indexList) {
+                                sharedBucketPriorityQueue.add(index.getTimestamp(), index.getLedgerId(),
+                                        index.getEntryId());
+                            }
+
+                            bucketState.setCurrentSegmentEntryId(nextSegmentEntryId);
+                        });
+            });
+        });
+    }
+
+    private void resetLastMutableBucketRange() {
+        lastMutableBucketState.setStartLedgerId(-1L);
+        lastMutableBucketState.setEndLedgerId(-1L);
+    }
+
+    @Override
+    public synchronized boolean addMessage(long ledgerId, long entryId, long deliverAt) {
+        if (containsMessage(ledgerId, entryId)) {
+            messagesHaveFixedDelay = false;
+            return true;
+        }
+
+        if (deliverAt < 0 || deliverAt <= getCutoffTime()) {
+            messagesHaveFixedDelay = false;
+            return false;
+        }
+
+        boolean existBucket = findBucket(ledgerId).isPresent();
+
+        // Create bucket snapshot
+        if (ledgerId > lastMutableBucketState.endLedgerId && !getPriorityQueue().isEmpty()) {
+            if (getPriorityQueue().size() >= minIndexCountPerBucket || existBucket) {
+                asyncCreateBucketSnapshot();
```

Comment:
The `addMessage` will be synchronous by synchronized, and all possible conflict operations in `asyncCreateBucketSnapshot` also are synchronous.

Reply:
I understand what you mean, `asyncCreateBucketSnapshot` is not a completely asynchronous method

1. build bucket synchronously ( in synchronized block )
2. persist asynchronously

Should we change its name to `createBucketSnapshotAndAsyncPersistent`? Because two people have already misunderstood this method.